### ISSUE TITLE
[FEATURE] Add test mode during install, RFC

### ISSUE THREAD

**Issue by nothingismagick** [quasarframework/quasar-starter-kit#43]
This feature is about enabling the installation and integration of unit test-runners, e2e frameworks and quality control into the Quasar Framework. This is how it should work:

0. The user calls `quasar init mypackage`
1. The user is asked if they want to include one of each of test types
 a. UNIT TEST (ava, jasmine, jest, mocha, tap) includes istanbul/nyc, chai, sinon 
 b. E2E TEST (cypress, karma, nightwatch)
 c. QUALITY (lighthouse)
2. These decisions are passed back to `quasar-cli`
3. `quasar-cli` adds the **`mypackage/test`** folder, installs the base `@quasar-test/base` repo 
4. Each of the above test runners is a folder within the namespaced monorepo which are also installed (such as `@quasar-test/ava`, `@quasar-test/cypress`, `@quasar-test/lighthouse`)
5. As a post-install step, test-runner configs, src and output folders are copied or created in the **`mypackage/test`** folder.
6. Where necessary, the **`mypackage/package.json`** is updated with configs for the test runners.
7. Test commands are created in the **`mypackage/package.json`**, so that the user can run tests with yarn/npm.
8. Where necessary, webpack-chain commands and aliases are added to **`mypackage/quasar.conf.js`**
9. **`mypackage/.babelrc`** is updated accordingly
10. Upon conclusion, the "test mode" is set and visible when the user runs `quasar mode`:
> ![screen shot 2018-04-17 at 15 34 41](https://user-images.githubusercontent.com/35242872/40575364-066690f6-60e4-11e8-87eb-4f7a0af7bfd2.png)

**Comment by nothingismagick** [comment]
I should note, that at the moment, the only test runners that are verified to have valid "recipes" at the moment are:

- ava
- jest
- cypress
- lighthouse

**Comment by Arushi-Jain** [comment]
In order to download testing dependencies along with the quasar template, I forked the quasar-starter-kit and changed the package.json. It downloads only the dependencies for unit test for now only when the user says yes to the question asked (edit meta.js to change the questions asked).
Quasar-cli is needed only to create the folder for those tests, for now I create the folder myself for unit testing while the dependencies are downloaded only i choose to. 
this is my meta.js


const
  path = require('path'),
  fs = require('fs')

const {
  sortDependencies,
  installDependencies,
  runLintFix,
  printMessage,
} = require('./utils')

const pkg = require('./package.json')

module.exports = {
  helpers: {
    template_version() {
      return pkg.version
    }
  },

  prompts: {
    name: {
      type: 'string',
      required: true,
      message: 'Project name (internal usage for dev)',
    },
    productName: {
      type: 'string',
      required: true,
      message: 'Project product name (official name)',
      default: 'Quasar App'
    },
    description: {
      type: 'string',
      required: false,
      message: 'Project description',
      default: 'A Quasar Framework app',
    },
    author: {
      type: 'string',
      message: 'Author',
    },
    lint: {
      type: 'confirm',
      message: 'Use ESLint to lint your code?',
    },
    lintConfig: {
      when: 'lint',
      type: 'list',
      message: 'Pick an ESLint preset',
      choices: [
        {
          name: 'Standard (https://github.com/standard/standard)',
          value: 'standard',
          short: 'Standard',
        },
        {
          name: 'Airbnb (https://github.com/airbnb/javascript)',
          value: 'airbnb',
          short: 'Airbnb',
        },
        {
          name: 'none (configure it yourself)',
          value: 'none',
          short: 'none',
        }
      ]
    },
    cordovaId: {
      type: 'string',
      required: false,
      message: 'Cordova id (disregard if not building mobile apps)',
      default: 'org.cordova.quasar.app'
    },
    vuex: {
      type: 'confirm',
      message: 'Use Vuex? (recommended for complex apps/websites)',
    },
    axios: {
      type: 'confirm',
      message: 'Use Axios for Ajax calls?'
    },
    test : {
      type: 'confirm',
      message: 'Use Unit Testing for your code ?'
    },
    i18n: {
      type: 'confirm',
      message: 'Use Vue-i18n? (recommended if you support multiple languages)',
      default: false
    },
    ie: {
      type: 'confirm',
      message: 'Support IE11?',
      default: false
    },
    autoInstall: {
      type: 'list',
      message:
        'Should we run `npm install` for you after the project has been created? (recommended)',
      choices: [
        {
          name: 'Yes, use NPM',
          value: 'npm',
          short: 'NPM',
        },
        {
          name: 'Yes, use Yarn (recommended)',
          value: 'yarn',
          short: 'yarn',
        },
        {
          name: 'No, I will handle that myself',
          value: false,
          short: 'no',
        }
      ]
    }
  },
  filters: {
    '.eslintrc.js': 'lint',
    '.eslintignore': 'lint',
    '.stylintrc': 'lint',
    'src/store/**/*': 'vuex',
    'src/i18n/**/*': 'i18n',
    'src/plugins/i18n.js': 'i18n',
    'src/plugins/axios.js': 'axios'
  },
  complete: function(data, { chalk }) {
    const green = chalk.green

    sortDependencies(data, green)

    const cwd = path.join(process.cwd(), data.inPlace ? '' : data.destDirName)

    if (data.autoInstall) {
      installDependencies(cwd, data.autoInstall, green)
        .then(() => {
          return runLintFix(cwd, data, green)
        })
        .then(() => {
          printMessage(data, green)
        })
        .catch(e => {
          console.log(chalk.red('Error:'), e)
        })
    }
    else {
      printMessage(data, chalk)
    }
  }
} 


and this my package.json 

{
  "name": "{{ name }}",
  "version": "1.0.0",
  "description": "{{ description }}",
  "productName": "{{ productName }}",
  "cordovaId": "{{ cordovaId }}",
  "author": "{{ author }}",
  "private": true,
  "scripts": {
    "converttonemf": "node build/convert.js" ,
    {{#lint}}"lint": "eslint --ext .js,.vue src --fix"{{/lint}}{{#test}}{{#lint}},{{/lint}}
    "test": "mocha-webpack --webpack-config webpack.test.conf.js test --recursive --require test/setup.js"{{/test}}
  },
  "dependencies": {
    "flo-components": "0.0.3",
    "chart.js": "^2.7.2"
    {{#axios}},{{/axios}}
    {{#axios}}"axios": "^0.18.0"{{/axios}}
    {{#i18n}},{{/i18n}}
    {{#i18n}}"vue-i18n": "^7.3.3"{{/i18n}}
  },
  "devDependencies": {
    {{#lint}}
    "babel-eslint": "^8.2.1",
    "eslint": "^4.18.2",
    "eslint-friendly-formatter": "^3.0.0",
    "eslint-loader": "^2.0.0",
    "eslint-plugin-vue": "^4.3.0",
    {{#if_eq lintConfig "standard"}}
    "eslint-config-standard": "^11.0.0",
    "eslint-plugin-promise": "^3.7.0",
    "eslint-plugin-standard": "^3.0.1",
    "eslint-plugin-import": "^2.9.0",
    "eslint-plugin-node": "^6.0.1",
    {{/if_eq}}
    {{#if_eq lintConfig "airbnb"}}
    "eslint-config-airbnb-base": "^12.1.0",
    "eslint-plugin-import": "^2.9.0",
    {{/if_eq}}
    {{/lint}}
    "quasar-cli": "^0.15.14" {{#test}},
    "@vue/test-utils": "^1.0.0-beta.15",
    "jsdom": "^11.9.0",
    "jsdom-global": "^3.0.2",
    "mocha": "^5.1.1",
    "mocha-webpack": "^1.1.0",
    "chai": "^4.1.2"
   {{/test}}
  },
  "engines": {
    "node": ">= 8.9.0",
    "npm": ">= 5.6.0"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "not ie <= 10"
  ]
}


Also i had to add webpack.test.config.js to the template folder to add the necessary webpack loaders and this config.js file is called only when we run the script "npm run test"

const path = require('path')

module.exports = {
    module: {
        loaders: [
            {
                test: /\.vue$/,
                loader: 'vue-loader'
            },
            {
                test: /\.js$/,
                loader: 'babel-loader',
                exclude: /node_modules/
            }
        ]
    },
    resolve: {
        extensions: ['.js', '.vue']
    }
}

You can refer to my forked repository for more

**Comment by nothingismagick** [comment]
@Arushi-Jain - thanks! can you do me a favour though, and remove the code from this RFC or hide the code blocks within \<details>\</details> ? And put it into an issue here: https://github.com/quasarframework/quasar-test

**Comment by nothingismagick** [comment]
Regarding your comment in general, we have moved to babel7 and webpack4 - your package.json shows that your source is not up to date with quasar 0.16.1...

**Comment by Arushi-Jain** [comment]
But this approach will even work, we are just creating our own template using quasar-starter-kit .
with quasar 0.16.1 we may have to change the version of the webpack loaders in webpack.test.config.js

**Comment by nothingismagick** [comment]
Yes, your approach would *theoretically* work, but the thing is that we want to centrally manage the packages installed for the various test runners. We are not merely offering mocha-webpack. Further, this approach does not hook into the webpack chain of quasar 16+ - which in my opinion is the only reason to use mocha webpack. Otherwise just use jest or Ava...

**Comment by nothingismagick** [comment]
For your information, I was also able to integrate mocha-webpack into quasar 15, it is Babel and the new webpack approach that makes it challenging with 16.

**Comment by Arushi-Jain** [comment]
Thank you for your insights, I will definitely look into it.

**Comment by TranDinhKhang** [comment]
why use mocha instead of jest?

Jest is very strong supported by facebook, we will need jest snapshot & image snapshot to get pixel perfect test for UI.
